djautotask/syncrest.py

```diff
@@ -0,0 +1,414 @@
+        if sync_job_qset.count() > 1 and not self.full:
+            last_sync_job_time = sync_job_qset.exclude(
+                id=sync_job_qset.last().id).last().start_time.isoformat()
+            self.api_conditions.append(
+                "lastUpdated>[{0}]".format(last_sync_job_time)
+            )
+        results = SyncResults()
+        results = self.get(results)
+
+        if self.full:
+            # Set of IDs of all records prior to sync,
+            # to find stale records for deletion.
+            initial_ids = self._instance_ids()
+
+            results.deleted_count = self.prune_stale_records(
+                initial_ids, results.synced_ids
+            )
+
+        return results.created_count, results.updated_count, \
+            results.skipped_count, results.deleted_count
+
+    def callback_sync(self, filter_params):
+
+        results = SyncResults()
+
+        # Set of IDs of all records related to the parent object
+        # to sync, to find stale records for deletion, need to be careful
+        # to get the correct filter on the objects you want, or you will
+        # be deleting records you didn't intend to, and they wont be restored
+        # until their next scheduled sync
+        initial_ids = self._instance_ids(filter_params=filter_params)
+        results = self.get(results)
+
+        # This should always be a full sync (unless something changes in
+        # the future and it doesn't need to delete anything)
+        results.deleted_count = self.prune_stale_records(
+            initial_ids, results.synced_ids
+        )
+
+        return results.created_count, results.updated_count, \
+            results.skipped_count, results.deleted_count
+
+    def sync_children(self, *args):
+        for synchronizer, filter_params in args:
+            created_count, updated_count, skipped_count, deleted_count \
+                = synchronizer.callback_sync(filter_params)
+            msg = '{} Child Sync - Created: {},'\
+                ' Updated: {}, Skipped: {},Deleted: {}'.format(
+                    synchronizer.model_class.__bases__[0].__name__,
+                    created_count,
+                    updated_count,
+                    skipped_count,
+                    deleted_count
+                )
+            logger.info(msg)
+
+    def remove_null_characters(self, json_data):
+        for value in json_data:
+            if isinstance(json_data.get(value), str):
+                json_data[value] = json_data[value].replace('\x00', '')
+
+        return json_data
+
+
+class ContactSynchronizer(Synchronizer):
+    client_class = api.ContactsAPIClient
+    model_class = models.ContactTracker
+
+    related_meta = {
+        'companyID': (models.Account, 'account'),
+    }
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.api_conditions = ['inactiveFlag=False']
+
+    def _assign_field_data(self, instance, json_data):
```

Comment:
Right now whenever `atsyncrest` is run any records that are returned are saved to the database. What we want is to run a sync and for each record returned check our database to see if we have the record, if we have it then update it, and if not save it as a new record.

We need to make sure that we set the `id` on the instance to the `id` that we get from the AT API. Something like `instance.id = json_data.get('id')`. This will only the synchronizer's to be able to find out if the record from the API is in our database.